src/lib.rs

```diff
@@ -256,55 +261,105 @@ impl Factorization {
             return Some(n)
         }
 
+        let org = n.clone().try_into().unwrap();
+        let mut n: BigInt = n.try_into().unwrap();
+
+        let mut f = BigInt::zero();
+
+        let one = BigInt::one();
+        let two: BigInt = &one + &one;
+
+        // This while loop will run until n is completely factorized,
+        // thus giving the largest factor. If you only care about A factor,
+        // this can be rewritten as below for much better performance.
+        
+        while n > one {
+        // while f == BigInt::zero() {
+
+
+            let mut x = two.clone();
+            let mut y = two.clone();
+            let mut d = one.clone();
+            let n_int = n.clone().try_into().unwrap();
+
+            let g = |x: &BigInt| (x.modpow(&two, &n_int) + 1) % &n_int;
+
+            while d == one {
+                x = g(&x);
+                y = g(&g(&y));
+                d = gcd((&x - &y).abs(), &n_int - 0);
+            }
+
+            if &d == &org {
+                return Some(brute_force_prime_factor(d.try_into().unwrap()));
+            }
+
+            n /= &d;
+
+            if &d > &f {
+                f = d;
+            }
+        }
+
+        return Some(f.try_into().unwrap());
+    }
+}
+
+fn brute_force_prime_factor(mut n: BigUint) -> BigUint {
+
         let one = BigUint::one();
         let two = &one + &one;
-        
+
         // STEP 1 | n divided by 2
         while n.is_even() {
             n = n / &two;
         }
-        
+
         // STEP 2 | 3..sqrt(n) | Divide i by n. On failure, add 2 to i
         let n_sqrt = n.sqrt().to_usize().unwrap();
         
-        for mut i in 3..n_sqrt {
-            while n.divides(&BigUint::from(i)) {
-                n = n / BigUint::from(i);
+        for mut i in 3..n_sqrt + 1 {
```

Comment:
We want to check the n_sqrt as well in this case since it is possible that n = p^2